lib/update/automatic.go

```diff
@@ -67,7 +68,8 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
 	force := false
 	fsmErr := fsm.ExecutePlan(ctx, progress, force)
 	if fsmErr != nil {
-		return trace.Wrap(err)
+		log.Warnf("Failed to execute plan: %v.", fsmErr)
```

Comment:
Hmm... I'm a bit confused by the logic of the fallthrough here. 

If I understand this, if Executing the plan has an error of some sort, we keep that error and call fsm.Complete. In fsm.Complete, it checks if the plan has been completed, and if not it tell's the ops.Service to Fail the Operation and mark it as complete.

Won't this basically the leave the cluster in a partially upgraded state, but also having released the cluster lock that prevents other operations? If someone retries the upgrade, will it try and continue the existing upgrade, or start a new plan?

Sorry, not sure if I'm just missing something here, without spending the time to dig through the operation code.

Reply:
Yes, you're right. I was thinking about the automatic update observed by the robotest. But it does not make sense to remove the cluster lock for upgrades - I'll have this fixed in the robotest code.

Reply:
On the other hand, I think I did not implement the `Complete` operation correctly - in case of error, it should mark only the operation as `failed` and leave the cluster in the state it was during the operation. This will prevent another upgrade operation.